### PR TITLE
Add the relevance manager module

### PR DESCRIPTION
**Pull request by ajreynol**
This PR adds the relevance manager module, which will be used in forthcoming PRs to query assignments for whether a literal is "relevant" (e.g. critical for satisfying the input formula).  Leveraging this technique has led to noticeable improvements for non-linear arithmetic (https://github.com/ajreynol/CVC4/tree/relManager).

This PR does not enable it, but adds the module only.

### CODE REVIEW

File: src/theory/relevance_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  if (k == NOT)
+  {
+    cache[cur] = -lastChildJustify;
+  }
+  else if (k == IMPLIES || k == AND || k == OR)
+  {
+    if (lastChildJustify != 0)
+    {
+      // see if we short circuited?
+      if (lastChildJustify
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:139]: Why don't we short circuit for `OR`?

File: src/theory/relevance_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+      cache[cur] = lastChildJustify;
+    }
+  }
+  else
+  {
+    Assert(k == XOR || k == EQUAL);
+    Assert(lastChildJustify != 0);
+    if (index == 0)
+    {
+      // must compute the other child
```
Comment (nafur) [inline src/theory/relevance_manager.cpp:200]: Both `XOR` and `EQUAL` can be n-ary in SMT-LIB. Can the `Node` be n-ary as well?

File: src/theory/relevance_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      // an "always relevant" set here.
+      d_input.push_back(a);
+    }
+    i++;
+  }
+}
+
+void RelevanceManager::notifyPreprocessedAssertion(Node n)
+{
+  d_input.push_back(n);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:72]: We can't, I'll fix. Notice this is only an optimization.

File: src/theory/relevance_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    }
+  }
+  Trace("rel-manager") << "...success, size = " << d_rset.size() << std::endl;
+  d_success = true;
+}
+
+bool RelevanceManager::isBooleanConnective(TNode cur)
+{
+  Kind k = cur.getKind();
+  return k == NOT || k == IMPLIES || k == AND || k == OR || k == ITE || k == XOR
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:108]: No, we don't traverse into term positions. I can add an assertion for this.

File: src/theory/relevance_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+}
+
+void RelevanceManager::computeRelevance()
+{
+  d_computed = true;
+  Trace("rel-manager") << "RelevanceManager::computeRelevance..." << std::endl;
+  std::unordered_map<TNode, int, TNodeHashFunction> cache;
+  for (NodeList::const_iterator it = d_input.begin(); it != d_input.end(); ++it)
+  {
+    TNode n = *it;
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:88]: I don't think range-based for loops work with context-dependent lists (d_input is `context::CDList<Node>`).

File: src/theory/relevance_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+  std::unordered_map<TNode, std::vector<int>, TNodeHashFunction> childJustify;
+  std::unordered_map<TNode, int, TNodeHashFunction>::iterator it;
+  std::unordered_map<TNode, std::vector<int>, TNodeHashFunction>::iterator itc;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    if (it != cache.end())
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:229]: Good catch, I had made some last minute changes and dropped that.

File: src/theory/relevance_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+      continue;
+    }
+    itc = childJustify.find(cur);
+    // have we traversed to children yet?
+    if (itc == childJustify.end())
+    {
+      // are we not a Boolean connective (including NOT)?
+      if (isBooleanConnective(cur))
+      {
+        childJustify[cur].clear();
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:242]: This is the line that initializes childJustify[cur] as empty.  Traditionally I do this via calling clear to "touch" it.

File: src/theory/relevance_manager.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:139]: This is short circuiting all cases, the code is captures the cases which "false" short circuits, but both true and false are being tested.

File: src/theory/relevance_manager.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/relevance_manager.cpp:200]: We normalize them (all the way at the API level).  I'll add an assertion.

File: src/theory/relevance_manager.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (nafur) [reply src/theory/relevance_manager.cpp:95]: range-based for loops should for for any object that implements `.begin()` and `.end()` (or `begin(obj)` and `end(obj)`, for that matter).

File: src/theory/relevance_manager.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nafur) [reply src/theory/relevance_manager.cpp:255]: Hm, okay. I guess it doesn't hurt.

**Pull request merged**